allennlp/semparse/type_declarations/wikitables_variable_free.py

```diff
@@ -14,43 +15,49 @@
 # cells, and the executor can process them accordingly, but the grammar does not differentiate
 # between them. Also note that columns are basic types in this grammar.
 ROW_TYPE = NamedBasicType("ROW")
-# TODO(pradeep): Add different column types for string, number, and date columns, and may be a
-# generic column type.
-# "LCOLUMN" to ensure the signature will be "l", to avoid confusion with cell. I decided to omit the
-# cell type eventually, and the decision to call this "LCOLUMN" is not relevant any more. But I'm
-# not changing it for now, because I might end up adding the cell type back later.
-COLUMN_TYPE = NamedBasicType("LCOLUMN")
+# The following three type signatures are assigned such that they're easy to understand from the
+# first letters, while being different from the cell types.
+DATE_COLUMN_TYPE = NamedBasicType("YCOLUMN")  # Y for year
+NUMBER_COLUMN_TYPE = NamedBasicType("FCOLUMN")  # F for float
+STRING_COLUMN_TYPE = NamedBasicType("TCOLUMN")  # T for token
+GENERIC_COLUMN_TYPE = MultiMatchNamedBasicType("GCOLUMN", [STRING_COLUMN_TYPE, DATE_COLUMN_TYPE,
+                                                           NUMBER_COLUMN_TYPE])
+COMPARABLE_COLUMN_TYPE = MultiMatchNamedBasicType("CCOLUMN", [NUMBER_COLUMN_TYPE, DATE_COLUMN_TYPE])
+
 NUMBER_TYPE = NamedBasicType("NUMBER")
 DATE_TYPE = NamedBasicType("DATE")
 STRING_TYPE = NamedBasicType("STRING")
 
-BASIC_TYPES = {ROW_TYPE, COLUMN_TYPE, NUMBER_TYPE, DATE_TYPE, STRING_TYPE}
+BASIC_TYPES = {ROW_TYPE, GENERIC_COLUMN_TYPE, COMPARABLE_COLUMN_TYPE, NUMBER_COLUMN_TYPE,
+               STRING_COLUMN_TYPE, DATE_COLUMN_TYPE, NUMBER_TYPE, DATE_TYPE, STRING_TYPE}
 STARTING_TYPES = {NUMBER_TYPE, DATE_TYPE, STRING_TYPE}
 
 # Complex types
 # Type for selecting the value in a column in a set of rows. "select" and "mode" functions.
-SELECT_TYPE = ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE, STRING_TYPE))
+SELECT_TYPE = ComplexType(ROW_TYPE, ComplexType(GENERIC_COLUMN_TYPE, STRING_TYPE))
 
 # Type for filtering rows given a column. "argmax", "argmin" and "same_as" (select all rows with the
-# same value under the given column as the given row does under the given column)
+# same value under the given column as the given row does under the given column). While "same_as"
+# takes any column, "argmax" and "argmin" take only comparable columns (i.e. dates or numbers).
 # Note that the values used for comparison in "argmax" and "argmin" can only come from column
 # lookups in this language. In LambdaDCS, there's a lambda function that is applied to the rows to
 # get the values, but here, we simply have a column name.
-ROW_FILTER_WITH_COLUMN = ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE, ROW_TYPE))
+ROW_FILTER_WITH_GENERIC_COLUMN = ComplexType(ROW_TYPE, ComplexType(GENERIC_COLUMN_TYPE, ROW_TYPE))
+ROW_FILTER_WITH_COMPARABLE_COLUMN = ComplexType(ROW_TYPE, ComplexType(COMPARABLE_COLUMN_TYPE, ROW_TYPE))
```

Comment:
It might be better to handle this function like we handle `UnaryOpTypes`, so that we don't have to add another type to account for this.  I guess this is the simpler way to go - you just need to add number and date columns to three types this way, instead of two - and given that we're thinking of building type inference on top of python, anyway, it's probably not worth the effort to fix up this type system.